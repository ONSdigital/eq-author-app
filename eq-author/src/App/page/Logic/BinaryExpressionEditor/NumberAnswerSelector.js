import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { get, some } from "lodash";

import { colors, radius } from "constants/theme";
import { Number, Select, Label } from "components/Forms";
import VisuallyHidden from "components/VisuallyHidden";

import {
  rightSideErrors,
  OPERATOR_REQUIRED,
} from "constants/validationMessages";
import ValidationError from "components/ValidationError";

const conditions = {
  SELECT: null,
  EQUAL: "Equal",
  NOT_EQUAL: "NotEqual",
  GREATER_THAN: "GreaterThan",
  LESS_THAN: "LessThan",
  GREATER_OR_EQUAL: "GreaterOrEqual",
  LESS_OR_EQUAL: "LessOrEqual",
  UNANSWERED: "Unanswered",
};

export const ConditionSelector = styled(Select)`
  width: auto;
`;

const Value = styled.div`
  flex: 1 1 auto;
  display: flex;
  margin-left: 1em;
  align-items: center;
  position: relative;
`;

const NumberAnswerRoutingSelector = styled.div`
  display: flex;
  align-items: center;
  border: 1px solid ${colors.lightGrey};
  margin: 1em 0;
  border-radius: ${radius};
  padding: 1em;
  ${({ hasError }) =>
    hasError &&
    `
    border-color: ${colors.errorPrimary};
    outline-color: ${colors.errorPrimary};
    box-shadow: 0 0 0 2px ${colors.errorPrimary};
    border-radius: 4px;
    margin-bottom: 0.5em;
  `}
`;

class NumberAnswerSelector extends React.Component {
  static propTypes = {
    expression: PropTypes.shape({
      id: PropTypes.string.isRequired,
      left: PropTypes.shape({
        type: PropTypes.string.isRequired,
      }).isRequired,
      right: PropTypes.shape({
        number: PropTypes.number,
      }),
    }).isRequired,
    onRightChange: PropTypes.func.isRequired,
    onConditionChange: PropTypes.func.isRequired,
    groupErrorMessage: PropTypes.string,
  };

  state = {
    number: get(this.props.expression, "right.number", null),
  };

  handleRightChange = ({ value }) => this.setState(() => ({ number: value }));

  handleRightBlur = () => {
    this.props.onRightChange({ customValue: { number: this.state.number } });
  };

  handleConditionChange = ({ value }) => {
    this.props.onConditionChange(value);
  };

  handleError = () => {
    const { expression, groupErrorMessage } = this.props;
    let message = null;

    const errors = expression.validationErrorInfo.errors;

    if (errors.some(({ field }) => field === "condition")) {
      message = OPERATOR_REQUIRED;
    } else if (
      errors.some(
        ({ errorCode }) =>
          errorCode === rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.errorCode
      )
    ) {
      message = rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.message;
    }

    if (
      some(expression.validationErrorInfo.errors, {
        errorCode: rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.errorCode,
      })
    ) {
      message = rightSideErrors.ERR_RIGHTSIDE_NO_VALUE.message;
    }

    return <ValidationError>{message || groupErrorMessage}</ValidationError>;
  };

  render() {
    const { expression, groupErrorMessage } = this.props;

    const hasError =
      expression.validationErrorInfo.errors.length > 0 || groupErrorMessage;

    return (
      <>
        <NumberAnswerRoutingSelector hasError={hasError}>
          <VisuallyHidden>
            <Label htmlFor={`expression-condition-${expression.id}`}>
              Operator
            </Label>
          </VisuallyHidden>
          <ConditionSelector
            id={`expression-condition-${expression.id}`}
            onChange={this.handleConditionChange}
            name="condition-select"
            value={expression.condition}
            data-test="condition-selector"
          >
            {!expression.condition && (
              <option value={conditions.SELECT}>Select an operator</option>
            )}
            <option value={conditions.EQUAL}>(=) Equal to</option>
            <option value={conditions.NOT_EQUAL}>(&ne;) Not equal to</option>
            <option value={conditions.GREATER_THAN}>(&gt;) More than</option>
            <option value={conditions.LESS_THAN}>(&lt;) Less than</option>
            <option value={conditions.GREATER_OR_EQUAL}>
              (&ge;) More than or equal to
            </option>
            <option value={conditions.LESS_OR_EQUAL}>
              (&le;) Less than or equal to
            </option>
            <option value={conditions.UNANSWERED}>Unanswered</option>
          </ConditionSelector>
          {expression.condition !== conditions.UNANSWERED &&
            expression.condition !== conditions.SELECT && (
              <>
                <Value>
                  <VisuallyHidden>
                    <Label htmlFor={`expression-right-${expression.id}`}>
                      Value
                    </Label>
                  </VisuallyHidden>
                  <Number
                    default={null}
                    id={`expression-right-${expression.id}`}
                    min={-99999999}
                    max={999999999}
                    placeholder="Value"
                    value={this.state.number}
                    name={`expression-right-${expression.id}`}
                    onChange={this.handleRightChange}
                    onBlur={this.handleRightBlur}
                    data-test="number-value-input"
                    type={expression.left.type}
                    unit={get(expression.left, "properties.unit", null)}
                  />
                </Value>
              </>
            )}
        </NumberAnswerRoutingSelector>
        {hasError && this.handleError()}
      </>
    );
  }
}

export default NumberAnswerSelector;
