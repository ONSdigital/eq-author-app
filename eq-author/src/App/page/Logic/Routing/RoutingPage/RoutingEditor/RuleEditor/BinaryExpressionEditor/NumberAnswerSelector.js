import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { get } from "lodash";

import { colors, radius } from "constants/theme";
import { Number, Select, Label } from "components/Forms";
import VisuallyHidden from "components/VisuallyHidden";

const conditions = {
  EQUAL: "Equal",
  NOT_EQUAL: "NotEqual",
  GREATER_THAN: "GreaterThan",
  LESS_THAN: "LessThan",
  GREATER_OR_EQUAL: "GreaterOrEqual",
  LESS_OR_EQUAL: "LessOrEqual",
  UNANSWERED: "Unanswered",
};

export const ConditionSelector = styled(Select)`
  width: auto;
`;

const Value = styled.div`
  flex: 1 1 auto;
  display: flex;
  margin-left: 1em;
  align-items: center;
  position: relative;
`;

const NumberAnswerRoutingSelector = styled.div`
  display: flex;
  align-items: center;
  border: 1px solid ${colors.lightGrey};
  margin: 1em 0;
  border-radius: ${radius};
  padding: 1em;
`;

class NumberAnswerSelector extends React.Component {
  static propTypes = {
    expression: PropTypes.shape({
      id: PropTypes.string.isRequired,
      left: PropTypes.shape({
        type: PropTypes.string.isRequired,
      }).isRequired,
      right: PropTypes.shape({
        number: PropTypes.number.isRequired,
      }),
    }).isRequired,
    onRightChange: PropTypes.func.isRequired,
    onConditionChange: PropTypes.func.isRequired,
  };

  state = {
    number: get(this.props.expression, "right.number", null),
  };

  handleRightChange = ({ value }) => this.setState(() => ({ number: value }));

  handleRightBlur = () => {
    this.props.onRightChange({ customValue: { number: this.state.number } });
  };

  handleConditionChange = ({ value }) => {
    this.props.onConditionChange(value);
  };

  render() {
    const { expression } = this.props;
    return (
      <NumberAnswerRoutingSelector>
        <VisuallyHidden>
          <Label htmlFor={`expression-condition-${expression.id}`}>
            Operator
          </Label>
        </VisuallyHidden>
        <ConditionSelector
          id={`expression-condition-${expression.id}`}
          onChange={this.handleConditionChange}
          name="condition-select"
          value={expression.condition}
          data-test="condition-selector"
        >
          <option value={conditions.EQUAL}>(=) Equal to</option>
          <option value={conditions.NOT_EQUAL}>(&ne;) Not equal to</option>
          <option value={conditions.GREATER_THAN}>(&gt;) More than</option>
          <option value={conditions.LESS_THAN}>(&lt;) Less than</option>
          <option value={conditions.GREATER_OR_EQUAL}>
            (&ge;) More than or equal to
          </option>
          <option value={conditions.LESS_OR_EQUAL}>
            (&le;) Less than or equal to
          </option>
          <option value={conditions.UNANSWERED}>Unanswered</option>
        </ConditionSelector>
        {expression.condition !== conditions.UNANSWERED && (
          <>
            <Value>
              <VisuallyHidden>
                <Label htmlFor={`expression-right-${expression.id}`}>
                  Value
                </Label>
              </VisuallyHidden>
              <Number
                id={`expression-right-${expression.id}`}
                min={-99999999}
                max={999999999}
                placeholder="Value"
                value={this.state.number}
                name={`expression-right-${expression.id}`}
                onChange={this.handleRightChange}
                onBlur={this.handleRightBlur}
                data-test="number-value-input"
                type={expression.left.type}
                unit={get(expression.left, "properties.unit", null)}
              />
            </Value>
          </>
        )}
      </NumberAnswerRoutingSelector>
    );
  }
}

export default NumberAnswerSelector;
