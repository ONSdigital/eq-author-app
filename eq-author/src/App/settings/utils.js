export const getThemeSettingsErrorCount = (questionnaire) =>
  questionnaire?.themeSettings?.validationErrorInfo?.totalCount +
    questionnaire?.validationErrorInfo?.totalCount || 0;
