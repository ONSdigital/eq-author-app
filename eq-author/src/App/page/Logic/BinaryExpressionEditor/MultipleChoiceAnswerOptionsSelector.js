import React from "react";
import styled from "styled-components";
import { includes, get, find } from "lodash";
import { PropTypes, string } from "prop-types";
import { TransitionGroup } from "react-transition-group";

import CheckboxChip from "./CheckboxChip";
import CheckboxChipTransition from "./CheckboxChipTransition";
import CheckboxOptionPicker from "./CheckboxOptionPicker";
import Popover from "./CheckboxSelectorPopup";
import ValidationError from "components/ValidationError";

import {
  rightSideErrors,
  OPERATOR_REQUIRED,
} from "constants/validationMessages";
import { colors } from "constants/theme";
import { RADIO } from "constants/answer-types";
import { Select } from "components/Forms";

import TextButton from "components/buttons/TextButton";
import ToggleChip from "components/buttons/ToggleChip";
import SecondaryConditionSelector from "./SecondaryConditionSelector";
import { enableOn } from "utils/featureFlags";

const answerConditions = {
  UNANSWERED: "Unanswered",
  COUNTOF: "CountOf",
  ALLOF: "AllOf",
  ANYOF: "AnyOf",
  ONEOF: "OneOf",
  NOTANYOF: "NotAnyOf",
};

const MultipleChoiceAnswerOptions = styled.div`
  align-items: center;
  display: inline-flex;
  flex-flow: row wrap;
  width: 100%;
  margin-top: 0.75em;
  margin-bottom: 1em;
  padding: 0.5em;
  border-radius: 4px;
  border: 2px solid ${colors.lighterGrey};
  ${({ hasError }) =>
    hasError &&
    `
    border-color: ${colors.errorPrimary};
    outline-color: ${colors.errorPrimary};
    box-shadow: 0 0 0 2px ${colors.errorPrimary};
    border-radius: 4px;
    margin-bottom: 0.5em;
  `}
`;

const Label = styled.label`
  font-size: 1em;
  font-weight: bold;
  color: ${colors.textLight};
`;

const ConditionSelect = styled(Select)`
  display: inline-block;
  width: auto;
  margin: 0 0.5em;
  padding: 0.3em 2em 0.3em 0.5em;
  ${({ hasError }) => hasError && `border: thin solid #d0021b`}
`;

const ChooseButton = styled(TextButton)`
  font-size: 1rem;
  background: ${colors.white};
  padding: 0.3em 0.5em;
  border-radius: 4px;
  color: ${colors.primary};
  border: 1px solid ${colors.primary};
  display: inline-flex;
  align-items: center;
  margin: 0.25rem;
  max-width: 12em;
  letter-spacing: 0;

  &:hover {
    color: ${colors.white};
    background: ${colors.primary};
  }

  &:focus,
  &:active {
    outline-width: 0;
  }

  &:focus {
    box-shadow: 0 0 0 3px ${colors.tertiary};
    outline: none;
  }
`;

const SelectedOptions = styled.div`
  display: flex;
  align-items: center;
  flex-wrap: wrap;
`;

class MultipleChoiceAnswerOptionsSelector extends React.Component {
  constructor(props) {
    super(props);
    this.state = { showPopup: false };
  }

  static propTypes = {
    expression: PropTypes.shape({
      condition: string,
      left: PropTypes.shape({
        options: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string.isRequired,
            label: PropTypes.string,
          })
        ),
      }).isRequired,
      right: PropTypes.shape({
        options: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string.isRequired,
          })
        ),
      }),
    }).isRequired,
    onRightChange: PropTypes.func.isRequired,
    onConditionChange: PropTypes.func.isRequired,
    groupErrorMessage: PropTypes.string,
  };

  get selectedOptionIds() {
    return get(this.props.expression.right, "options", []).map(({ id }) => id);
  }

  handleRadioChange = ({ name: id, value: checked }) => {
    let newSelectedOptions;
    if (checked) {
      newSelectedOptions = [...this.selectedOptionIds, id];
    } else {
      newSelectedOptions = this.selectedOptionIds.filter(
        (selectedId) => selectedId !== id
      );
    }
    this.props.onRightChange({ selectedOptions: newSelectedOptions });
  };

  handleCheckboxUnselect = (id) => {
    const { expression } = this.props;
    const selectedOptions = expression.right.options
      .filter((option) => option.id !== id)
      .map(({ id }) => id);
    this.props.onRightChange({ selectedOptions });
  };

  handleConditionChange = ({ value }) => {
    this.props.onConditionChange(value);
  };

  handlePickerClose = () => {
    this.setState({ showPopup: false });
  };

  handleError = () => {
    const { expression, groupErrorMessage } = this.props;
    let message = groupErrorMessage;

    const error = find(
      expression.validationErrorInfo.errors,
      (error) =>
        error.errorCode.includes("ERR_RIGHTSIDE") ||
        error.errorCode.includes(
          "ERR_GROUP_MIXING_EXPRESSIONS_WITH_OR_STND_OPTIONS_IN_AND"
        )
    );

    if (
      expression.validationErrorInfo.errors.some(
        ({ field }) => field === "secondaryCondition"
      )
    ) {
      message = OPERATOR_REQUIRED;
    } else if (error) {
      if (expression.condition === "CountOf") {
        message = rightSideErrors[error.errorCode].message;
      } else {
        message =
          message ||
          rightSideErrors[error.errorCode].optionsMessage ||
          rightSideErrors[error.errorCode].message;
      }
    }

    return message ? <ValidationError>{message}</ValidationError> : null;
  };

  renderRadioOptionSelector(hasError) {
    const { expression } = this.props;
    const options = get(expression, "left.options", []);

    return (
      <>
        <MultipleChoiceAnswerOptions
          data-test="options-selector"
          hasError={hasError}
        >
          <Label>Match</Label>
          <ConditionSelect
            onChange={this.handleConditionChange}
            defaultValue={expression.condition}
            data-test="condition-dropdown"
          >
            <option value={answerConditions.ONEOF}>One of</option>
            <option value={answerConditions.UNANSWERED}>Unanswered</option>
          </ConditionSelect>
          {expression.condition !== answerConditions.UNANSWERED &&
            options.map((option) => (
              <ToggleChip
                key={option.id}
                name={option.id}
                title={option.label}
                checked={includes(this.selectedOptionIds, option.id)}
                onChange={this.handleRadioChange}
              >
                {option.label || <strong>Unlabelled option</strong>}
              </ToggleChip>
            ))}
        </MultipleChoiceAnswerOptions>
        {this.handleError()}
      </>
    );
  }

  renderCheckboxOptionSelector(hasError, hasConditionError) {
    const { expression } = this.props;

    return (
      <>
        <MultipleChoiceAnswerOptions
          data-test="options-selector"
          hasError={hasError}
        >
          <Label>Match</Label>
          <ConditionSelect
            onChange={this.handleConditionChange}
            defaultValue={expression.condition}
            hasError={hasConditionError}
            data-test="condition-dropdown"
          >
            <option value={answerConditions.ANYOF}>Any of</option>
            <option value={answerConditions.NOTANYOF}>Not any of</option>
            <option value={answerConditions.ALLOF}>All of</option>
            {enableOn(["enableCountCondition"]) && (
              <option value={answerConditions.COUNTOF}>Count of</option>
            )}
            <option value={answerConditions.UNANSWERED}>Unanswered</option>
          </ConditionSelect>
          {expression.condition === answerConditions.COUNTOF &&
            enableOn(["enableCountCondition"]) && (
              <SecondaryConditionSelector
                expression={expression}
                onRightChange={this.props.onRightChange}
                onConditionChange={this.props.onConditionChange}
              />
            )}
          {expression.condition !== answerConditions.UNANSWERED &&
            expression.condition !== answerConditions.COUNTOF && (
              <>
                <TransitionGroup component={SelectedOptions}>
                  {get(expression, "right.options", []).map((option) => {
                    const isMutuallyExclusive =
                      expression.left.mutuallyExclusiveOption &&
                      option.id === expression.left.mutuallyExclusiveOption.id;

                    return (
                      <CheckboxChipTransition key={option.id}>
                        <CheckboxChip
                          key={option.id}
                          id={option.id}
                          onRemove={this.handleCheckboxUnselect}
                          isMutuallyExclusive={isMutuallyExclusive}
                        >
                          {option.label || <strong>Unlabelled option</strong>}
                        </CheckboxChip>
                      </CheckboxChipTransition>
                    );
                  })}
                </TransitionGroup>
                <ChooseButton
                  onClick={() => {
                    this.setState({
                      showPopup: true,
                    });
                  }}
                >
                  Choose
                </ChooseButton>
              </>
            )}
          {this.state.showPopup && (
            <Popover isOpen onClose={this.handlePickerClose}>
              <CheckboxOptionPicker
                expression={expression}
                onClose={this.handlePickerClose}
                onRightChange={this.props.onRightChange}
              />
            </Popover>
          )}
        </MultipleChoiceAnswerOptions>
        {this.handleError()}
      </>
    );
  }

  render() {
    const { expression, groupErrorMessage } = this.props;
    const answerType = get(expression, "left.type");

    const { errors } = expression.validationErrorInfo;

    const hasError = errors.length || groupErrorMessage;
    const hasConditionError =
      errors.filter(({ field }) => field === "condition").length > 0;

    if (answerType === RADIO) {
      return this.renderRadioOptionSelector(hasError);
    } else {
      return this.renderCheckboxOptionSelector(hasError, hasConditionError);
    }
  }
}

export default MultipleChoiceAnswerOptionsSelector;
