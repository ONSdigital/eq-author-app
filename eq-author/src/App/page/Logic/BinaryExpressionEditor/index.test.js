import React from "react";
import { shallow } from "enzyme";
import { render, flushPromises, act, screen } from "tests/utils/rtl";
import { useMutation } from "@apollo/react-hooks";

import { RADIO, CURRENCY, NUMBER, PERCENTAGE } from "constants/answer-types";
import {
  NO_ROUTABLE_ANSWER_ON_PAGE,
  SELECTED_ANSWER_DELETED,
  DEFAULT_ROUTING,
} from "constants/routing-left-side";

import { byTestAttr } from "tests/utils/selectors";

import { UnwrappedBinaryExpressionEditor as BinaryExpressionEditor } from ".";
import MultipleChoiceAnswerOptionsSelector from "./MultipleChoiceAnswerOptionsSelector";
import NumberAnswerSelector from "./NumberAnswerSelector";

import { binaryExpressionErrors } from "constants/validationMessages";

import { OR } from "constants/routingOperators";

jest.mock("@apollo/react-hooks", () => ({
  useMutation: jest.fn(() => [() => null]),
}));

describe("BinaryExpressionEditor", () => {
  let defaultProps, expression, expressionGroup;
  beforeEach(() => {
    expression = {
      id: "1",
      left: {
        id: "2",
        type: RADIO,
      },
      condition: "Equal",
      secondaryCondition: null,
      right: null,
      validationErrorInfo: {
        id: "6dd",
        errors: [],
        totalCount: 0,
      },
    };

    expressionGroup = {
      expressions: [expression],
    };

    expression.expressionGroup = {
      id: "1",
      validationErrorInfo: { id: "1", errors: [], totalCount: 0 },
      __typename: "ExpressionGroup2",
    };

    defaultProps = {
      deleteBinaryExpression: jest.fn(),
      updateLeftSide: jest.fn(),
      updateRightSide: jest.fn(),
      updateBinaryExpression: jest.fn(),
      createBinaryExpression: jest.fn(),
      isOnlyExpression: false,
      isLastExpression: false,
      expressionGroup,
      expression,
      expressionIndex: 0,
      canAddCondition: true,
      match: {
        params: {
          questionnaireId: "1",
          sectionId: "2",
          pageId: "3",
        },
      },
    };
  });

  it("should render consistently", () => {
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should disable the delete expression button when there's only one expression and conditionType is not skip", () => {
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    expect(
      wrapper.find(byTestAttr("btn-remove")).prop("disabled")
    ).toBeTruthy();
  });

  it("should not disable the delete expression button when there's only one expression and conditionType is skip", () => {
    defaultProps.conditionType = "skip";
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    expect(wrapper.find(byTestAttr("btn-remove")).prop("disabled")).toBeFalsy();
  });

  it("should render multiple choice editor correctly", () => {
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    expect(wrapper.find(MultipleChoiceAnswerOptionsSelector)).toBeTruthy();
  });

  it("should render number editor correctly", () => {
    const answerTypes = [CURRENCY, NUMBER, PERCENTAGE];
    answerTypes.forEach((answerType) => {
      defaultProps.expression.left.type = answerType;
      const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
      expect(wrapper.find(NumberAnswerSelector)).toBeTruthy();
    });
  });

  it("should call createBinaryExpression when add button is clicked", () => {
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    wrapper.find(byTestAttr("btn-add")).simulate("click");
    expect(defaultProps.createBinaryExpression).toHaveBeenCalledWith(
      defaultProps.expressionGroupId
    );
  });

  it("should call deleteBinaryExpression when remove button is clicked and expressions length is greater than 1", () => {
    // Pushes newExpression to expressions array to increase expressions length above 1
    const newExpression = {
      id: "2",
      left: {
        id: "3",
        type: RADIO,
      },
      condition: "Equal",
      secondaryCondition: null,
      right: null,
      validationErrorInfo: {
        id: "6dd",
        errors: [],
        totalCount: 0,
      },
    };
    defaultProps.expressionGroup.expressions.push({ newExpression });

    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    wrapper.find(byTestAttr("btn-remove")).simulate("click");
    expect(defaultProps.deleteBinaryExpression).toHaveBeenCalledWith(
      defaultProps.expression.id,
      expect.any(Function)
    );
  });

  it("should call deleteSkipCondition when remove button is clicked, expressions length is 1 and conditionType is skip", () => {
    const deleteSkipCondition = jest.fn();
    defaultProps.conditionType = "skip";
    useMutation.mockImplementationOnce(jest.fn(() => [deleteSkipCondition]));

    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    wrapper.find(byTestAttr("btn-remove")).simulate("click");
    expect(deleteSkipCondition).toHaveBeenCalledWith({
      variables: { input: { id: defaultProps.expressionGroup.id } },
    });
  });

  it("should correctly submit from RoutingAnswerContentPicker", () => {
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);

    wrapper.find(byTestAttr("routing-answer-picker")).simulate("submit", {
      value: {
        id: "999",
      },
    });
    expect(defaultProps.updateLeftSide).toHaveBeenCalledWith(
      defaultProps.expression,
      "999"
    );
  });

  it("should display the correct error message when there is no routable answer on page", async () => {
    defaultProps.expression.left = { reason: NO_ROUTABLE_ANSWER_ON_PAGE };

    render(<BinaryExpressionEditor {...defaultProps} />);

    await act(async () => {
      await flushPromises();
    });

    expect(
      screen.getByText(
        "No routable answers have been added to this question yet"
      )
    ).toBeTruthy();
  });

  it("should display the correct error message when the answer has been deleted", async () => {
    defaultProps.expression.left = { reason: SELECTED_ANSWER_DELETED };

    render(<BinaryExpressionEditor {...defaultProps} />);

    await act(async () => {
      await flushPromises();
    });

    expect(
      screen.getByText("The answer used in this condition has been deleted")
    ).toBeTruthy();
  });

  it("should display the correct error message when the routable answer has been moved after the routing Question", async () => {
    defaultProps.expression.validationErrorInfo.totalCount = 1;
    defaultProps.expression.validationErrorInfo.errors[0] = {
      errorCode: "ERR_LEFTSIDE_NO_LONGER_AVAILABLE",
      field: "left",
      id: "expression-routing-1-left",
      type: "expressions",
    };
    render(<BinaryExpressionEditor {...defaultProps} />);

    await act(async () => {
      await flushPromises();
    });

    expect(
      screen.getByText(
        "Select an answer that is not later on in the questionnaire"
      )
    ).toBeTruthy();
  });

  it("should display the correct error message when you can't add a second 'And' condition", async () => {
    render(
      <BinaryExpressionEditor {...defaultProps} canAddCondition={false} />
    );

    await act(async () => {
      await flushPromises();
    });

    expect(
      screen.getByText("AND condition not valid with 'radio button' answer")
    ).toBeTruthy();
  });
  it("should display the correct error message when you can't add a second 'Or' condition", async () => {
    defaultProps.expressionGroup.operator = OR;

    render(
      <BinaryExpressionEditor {...defaultProps} canAddCondition={false} />
    );

    await act(async () => {
      await flushPromises();
    });

    expect(
      screen.getByText(
        "OR condition is not valid when creating multiple radio rules"
      )
    ).toBeTruthy();
  });

  it("should update the binary expression when the condition is changed", () => {
    defaultProps.expression.left.type = CURRENCY;
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    wrapper
      .find(NumberAnswerSelector)
      .simulate("conditionChange", "NotEqual", null);
    expect(defaultProps.updateBinaryExpression).toHaveBeenCalledWith(
      defaultProps.expression,
      "NotEqual",
      null
    );
  });

  it("should update the right side when the right side is changed", () => {
    defaultProps.expression.left.type = NUMBER;
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    wrapper
      .find(NumberAnswerSelector)
      .simulate("rightChange", { customValue: { number: 123 } });
    expect(defaultProps.updateRightSide).toHaveBeenCalledWith(
      defaultProps.expression,
      {
        customValue: { number: 123 },
      }
    );
  });

  it("leftside should not be null when updating answer selector", () => {
    defaultProps.expression.left.type = NUMBER;
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    wrapper.find(NumberAnswerSelector);
    expect(defaultProps.updateLeftSide).not.toBeNull();
  });

  it("rightside should not be null when updating answer selector", () => {
    defaultProps.expression.left.type = NUMBER;
    const wrapper = shallow(<BinaryExpressionEditor {...defaultProps} />);
    wrapper.find(NumberAnswerSelector);
    expect(defaultProps.updateRightSide).not.toBeNull();
  });

  it("should not show the delete or add expression buttons when showing defaultRouting", async () => {
    render(<BinaryExpressionEditor {...defaultProps} />);

    await act(async () => {
      await flushPromises();
    });

    const actionBtns = screen.getByTestId("action-btns");
    expect(actionBtns).toHaveStyleRule("display: none;");
  });

  it("should not show the expression editor when showing defaultRouting", async () => {
    defaultProps.expression.left.reason = DEFAULT_ROUTING;
    render(<BinaryExpressionEditor {...defaultProps} />);

    await act(async () => {
      await flushPromises();
    });

    const operatorSelect = screen.queryByText("Select an operator");
    expect(operatorSelect).toBeFalsy();
  });

  it("should provide validation message when errors are present", async () => {
    defaultProps.expression.validationErrorInfo.totalCount = 1;
    defaultProps.expression.validationErrorInfo.errors[0] = {
      errorCode: "ERR_ANSWER_NOT_SELECTED",
      field: "left",
      id: "expression-routing-1-left",
      type: "expressions",
    };

    render(<BinaryExpressionEditor {...defaultProps} />);

    await act(async () => {
      await flushPromises();
    });

    expect(screen.getByTestId("action-btns")).toHaveStyleRule("display: flex;");
    expect(screen.getByText("Answer required")).toBeTruthy();
  });

  it("should pass on shared expressionGroup messages when group errors are present", async () => {
    defaultProps.expression.expressionGroup = {
      id: "1",
      validationErrorInfo: {
        id: "err-1",
        totalCount: 1,
        errors: [
          {
            errorCode: "ERR_LOGICAL_AND",
            field: "2",
            id: "exp_1",
            type: "expressions",
          },
        ],
      },
    };

    const { getByText } = render(<BinaryExpressionEditor {...defaultProps} />);

    await act(async () => {
      await flushPromises();
    });

    expect(getByText(binaryExpressionErrors.ERR_LOGICAL_AND)).toBeTruthy();
  });

  it("shouldn't pass on shared expressionGroup messages when group errors are present for a different answerId", async () => {
    defaultProps.expression.expressionGroup = {
      id: "1",
      validationErrorInfo: {
        id: "1",
        totalCount: 1,
        errors: [
          {
            errorCode: "ERR_LOGICAL_AND",
            field: "different_answer_id",
            id: "1",
            type: "expressions",
          },
        ],
      },
    };

    render(<BinaryExpressionEditor {...defaultProps} />);

    await act(async () => {
      await flushPromises();
    });

    expect(
      screen.queryByText(binaryExpressionErrors.ERR_LOGICAL_AND)
    ).toBeNull();
  });
});
