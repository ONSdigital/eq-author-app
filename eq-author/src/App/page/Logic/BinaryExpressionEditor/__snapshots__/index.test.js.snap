// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinaryExpressionEditor should render consistently 1`] = `
<Fragment>
  <Grid
    align="center"
    fillHeight={true}
  >
    <Column
      cols={1.5}
      gutters={false}
    >
      <components__StyledLabel
        data-test="routing-binary-expression-if-label"
        inline={true}
      >
        If
      </components__StyledLabel>
    </Column>
    <Column
      cols={8}
      gutters={false}
    >
      <RoutingAnswerContentPicker
        data-test="routing-answer-picker"
        hasError={false}
        onSubmit={[Function]}
        selectedId="2"
      />
    </Column>
    <Column
      cols={2.5}
      gutters={false}
    >
      <components__ActionButtons
        data-test="action-btns"
        horizontal={true}
      >
        <Tooltip
          content="Remove condition"
          place="top"
        >
          <components__RemoveButton
            data-test="btn-remove"
            disabled={true}
            onClick={[Function]}
            small={true}
            type="button"
            variant="primary"
          >
            <IconText
              hideText={true}
              icon={[Function]}
            >
              Remove condition
            </IconText>
          </components__RemoveButton>
        </Tooltip>
        <Tooltip
          content="Add condition"
          place="top"
        >
          <components__AddButton
            data-test="btn-add"
            disabled={false}
            onClick={[Function]}
            small={true}
            type="button"
            variant="primary"
          >
            <IconText
              hideText={true}
              icon={[Function]}
            >
              Add condition
            </IconText>
          </components__AddButton>
        </Tooltip>
      </components__ActionButtons>
    </Column>
  </Grid>
  <Grid
    align="top"
    fillHeight={true}
  >
    <Column
      cols={1.5}
      gutters={false}
    >
      <components__ConnectedPath
        pathEnd={true}
      />
    </Column>
    <Column
      cols={8}
      gutters={false}
    >
      <Transition
        in={true}
        mountOnEnter={true}
        timeout={400}
        unmountOnExit={true}
      >
        <MultipleChoiceAnswerOptionsSelector
          expression={
            Object {
              "condition": "Equal",
              "expressionGroup": Object {
                "__typename": "ExpressionGroup2",
                "id": "1",
                "validationErrorInfo": Object {
                  "errors": Array [],
                  "id": "1",
                  "totalCount": 0,
                },
              },
              "id": "1",
              "left": Object {
                "id": "2",
                "type": "Radio",
              },
              "right": null,
              "secondaryCondition": null,
              "validationErrorInfo": Object {
                "errors": Array [],
                "id": "6dd",
                "totalCount": 0,
              },
            }
          }
          onConditionChange={[Function]}
          onRightChange={[Function]}
        />
      </Transition>
    </Column>
    <Column
      cols={2.5}
    />
  </Grid>
  <Grid
    align="top"
    fillHeight={true}
  >
    <Column
      cols={1.5}
    />
    <Column
      cols={8}
    />
    <Column
      cols={2.5}
    />
  </Grid>
</Fragment>
`;
